lib/l10n_utils/__init__.py

```diff
@@ -144,56 +143,19 @@ def get_locale(request):
     return getattr(request, "locale", settings.LANGUAGE_CODE)
 
 
-def get_accept_languages(request):
-    """
-    Parse the user's Accept-Language HTTP header and return a list of languages
-    """
-    languages = []
-    pattern = re.compile(r"^([A-Za-z]{2,3})(?:-([A-Za-z]{2})(?:-[A-Za-z0-9]+)?)?$")
-
-    parsed = parse_accept_lang_header(request.META.get("HTTP_ACCEPT_LANGUAGE", ""))
-
-    for lang, priority in parsed:
-        m = pattern.match(lang)
-
-        if not m:
-            continue
-
-        lang = m.group(1).lower()
-
-        # Check if the shorter code is supported. This covers obsolete long
-        # codes like fr-FR (should match fr) or ja-JP (should match ja)
-        if m.group(2) and lang not in settings.PROD_LANGUAGES:
-            lang += "-" + m.group(2).upper()
-
-        if lang not in languages:
-            languages.append(lang)
-
-    return languages
-
-
 def get_best_translation(translations, accept_languages):
-    # Look for the user's Accept-Language HTTP header to find another
-    # locale we can offer
-    for lang in accept_languages:
-        if lang in translations:
-            return lang
-
-    # Check for the fallback locales if the previous look-up doesn't
-    # work. This is useful especially in the Spanish locale where es-ES
-    # should be offered as the fallback of es, es-AR, es-CL and es-MX
-    for lang in accept_languages:
-        lang = settings.FALLBACK_LOCALES.get(lang)
-        if lang in translations:
-            return lang
-
-    # If all the attempts failed, just use en-US, the default locale of
-    # the site
-    if settings.LANGUAGE_CODE in translations:
-        return settings.LANGUAGE_CODE
-
-    # fall back to just the first locale in the list
-    return translations[0]
+    ranked = parse_accept_lang_header(accept_languages)
+    lang_map = _get_language_map()
+    valid_lang_map = {k: v for k, v in lang_map.items() if v in translations}
+    for lang, _ in ranked:
+        lang.lower()
+        if lang in valid_lang_map:
+            return valid_lang_map[lang]
+        pre = lang.split("-")[0]
+        if pre in valid_lang_map:
+            return valid_lang_map[pre]
+
+    return lang_map[translations[0]]
```

Comment:
Re-reading this, since `translations` is the template languages I don't think this needs to run through the mapping and could just return the first one as a fall back.